File: microservices/user-service/src/main/java/com/halcyon/tinder/userservice/dto/error/ErrorDetailsResponse.java
package com.halcyon.tinder.userservice.dto.error;

import com.fasterxml.jackson.annotation.JsonProperty;
import java.time.Instant;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;

@Builder
@Data
@AllArgsConstructor
public class ErrorDetailsResponse {

    @JsonProperty("timestamp")
    private Instant timestamp;

    @JsonProperty("status")
    private Integer status;

    @JsonProperty("error")
    private String error;

    @JsonProperty("message")
    private String message;
}

File: common/exception-core/src/main/java/com/halcyon/tinder/exceptioncore/ErrorDetailsResponse.kt
package com.halcyon.tinder.exceptioncore

import com.fasterxml.jackson.annotation.JsonProperty
import java.time.Instant

data class ErrorDetailsResponse(
    @JsonProperty("timestamp")
    val timestamp: Instant = Instant.now(),

    @JsonProperty("status")
    val status: Int,

    @JsonProperty("error")
    val error: String,

    @JsonProperty("message")
    val message: String
) {
    companion object {
        @JvmStatic
        fun builder() = Builder()
    }

    class Builder {
        private var timestamp: Instant = Instant.now()
        private var status: Int = 0
        private var error: String = ""
        private var message: String = ""

        fun timestamp(timestamp: Instant) = apply { this.timestamp = timestamp }
        fun status(status: Int) = apply { this.status = status }
        fun error(error: String) = apply { this.error = error }
        fun message(message: String) = apply { this.message = message }
        fun build() = ErrorDetailsResponse(timestamp, status, error, message)
    }
}